app/controllers/settings/two_factor_authentication/webauthn_credentials_controller.rb

```diff
@@ -42,23 +41,45 @@ def create
 
           if user_credential.save
             flash[:success] = I18n.t('webauthn_credentials.create.success')
-            status = :ok
+
+            generated_otp_backup_codes = false
+            unless current_user.otp_backup_codes?
+              current_user.otp_secret = User.generate_otp_secret(32)
+              @recovery_codes = current_user.generate_otp_backup_codes!
+              current_user.save!
+
+              generated_otp_backup_codes = true
+            end
 
             if current_user.webauthn_credentials.size == 1
               UserMailer.webauthn_enabled(current_user).deliver_later!
             else
               UserMailer.webauthn_credential_added(current_user, user_credential).deliver_later!
             end
+
+            if generated_otp_backup_codes
+              render json: {
+                html_data:
+                  render_to_string(
+                    partial: 'settings/two_factor_authentication/webauthn_credentials/recovery_codes',
+                    formats: :html,
+                    layout: false
+                  ),
+                status: :ok,
+              }
```

Comment:
As now you can just have Webauthn as 2FA, we need to generate the recovery codes (if they weren't generated already) when adding a security key.

As recovery codes need to do be rendered when generated because you cannot access them later – they are stored in the database as bcrypt hashes – we need to render the partial here and return it as part of the `json` response.

We could avoid this by adding the ability to access the codes _after_ they were generated – [`devise-two-factor` supports this in order to be able to display them to the user anytime](https://github.com/tinfoil/devise-two-factor/issues/64) – but I didn't wanted to make that change as part of this PR in order to avoid increasing its complexity. I do think we can work on that on a following PR.